plugins/TauMVAProducer.cc

```diff
@@ -0,0 +1,371 @@
+
+#include "TLorentzVector.h"
+
+class TauMVAProducer : public edm::stream::EDProducer<> 
+{
+public:
+    explicit TauMVAProducer(const edm::ParameterSet&);
+    ~TauMVAProducer();
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+    bool  isA(int particleID, int pdgId, bool checkCharge = false);
+    int   getContainingJetIndex(const pat::PackedCandidate* pfc) const;
+    int   getNearPhotonIndex(const pat::PackedCandidate* pfc);
+    float getDRNearestTrack(const pat::PackedCandidate* particle, const float minTrackPt=1.0);
+    float computePFIsolation(const pat::PackedCandidate* particle, const float minDR, const float maxDR, const unsigned int isotype=0, const float minNeutralPt=0.5, const float minPUPt=0.5, const float dBeta=0.5);
+    float TrackIso(const pat::PackedCandidate* particle, const float maxDR=0.3, const float deltaZCut=0.1);
+     
+private:
+     
+    virtual void beginStream(edm::StreamID) override;
+    virtual void produce(edm::Event&, const edm::EventSetup&) override;
+    virtual void endStream() override;
+
+    // ----------member data ---------------------------
+    edm::EDGetTokenT<std::vector<pat::Jet> >             jetToken_;
+    edm::EDGetTokenT<std::vector<pat::PackedCandidate> > pfCandToken_;
+    double        candptMin_;
+    double        candetaMax_;
+};
+
+TauMVAProducer::TauMVAProducer(const edm::ParameterSet& iConfig)
+{
+    //now do what ever other initialization is needed
+    produces<nanoaod::FlatTable>("pfcands");
+    produces<edm::PtrVector<reco::Candidate> >();
+    edm::InputTag jetSrc_    = iConfig.getParameter<edm::InputTag>("jetsSrc");
+    edm::InputTag pfCandSrc_ = iConfig.getParameter<edm::InputTag>("pfCandSrc");
+
+    jetToken_    = consumes<std::vector<pat::Jet> >(jetSrc_);
+    pfCandToken_ = consumes<std::vector<pat::PackedCandidate> >(pfCandSrc_);
+
+    candptMin_ = iConfig.getParameter<double>("minCandPt");
+    candetaMax_ = iConfig.getParameter<double>("maxCandEta");
+}
+
+TauMVAProducer::~TauMVAProducer()
+{
+ 
+}
+
+//
+// member functions
+//
+
+bool TauMVAProducer::isA(int particleID, int pdgId, bool checkCharge)
+{
+  int id = pdgId;
+  if(checkCharge) return (id == particleID);
+
+  return (std::fabs(id) == particleID);
+}
+
+int TauMVAProducer::getContainingJetIndex(const pat::PackedCandidate* pfc) const
+{
+  edm::Handle<std::vector<pat::Jet> > jets_;
+
+  int index = -1;
+  
+  for (unsigned int ijet = 0; ijet < jets_->size(); ijet++) {
+    const pat::Jet &j = (*jets_)[ijet];
+    for(unsigned int id = 0; id < j.numberOfDaughters(); id++) {
+      const pat::PackedCandidate* dau = dynamic_cast<const pat::PackedCandidate*>(j.daughter(id));
+      if(pfc == dau) {
+        index = ijet;
+        break;
+      }
+    }
+  }
+
+
+  return index;
+
+}
+
+int TauMVAProducer::getNearPhotonIndex(const pat::PackedCandidate* pfc)
+{
+  edm::Handle<std::vector<pat::PackedCandidate> > pfcands_;
```

Comment:
Mm.. is this the old code? How this really works?

Reply:
didn't you pass pfc by argument? But it is not used in the code

Reply:
It is used in line 146. This is a loop to find the nearest photon to the input packedCandidate

Reply:
OK， but pfcands_ is an empty handle, without getbyToken for example. So this code won't work. You can just pass the pfcands_ as argument from the producer()